ping/_compositions/rust-cross-versions.toml

```diff
@@ -0,0 +1,53 @@
+[metadata]
+  name = "rust-cross-versions {{ .Env.GitReference }}"
+
+[global]
+  plan = "libp2p/ping/rust"
+  case = "ping"
+  builder = "docker:generic"
+  total_instances = {{ if .Env.GitReference }}5{{ else }}4{{ end }}
+  runner = "local:docker"
+
+[[groups]]
+  id = "v0.44.0"
+  instances = { count = 1 }
+
+  [groups.build_config.build_args]
+    CARGO_FEATURES = 'v044'
+
+[[groups]]
+  id = "v0.45.1"
+  instances = { count = 1 }
+
+  [groups.build_config.build_args]
+    CARGO_FEATURES = 'v045'
+
+[[groups]]
+  id = "v0.46.0"
+  instances = { count = 1 }
+
+  [groups.build_config.build_args]
+    CARGO_FEATURES = 'v046'
+
+[[groups]]
+  id = "master"
+  instances = { count = 1 }
+
+  [groups.build_config.build_args]
+    CARGO_FEATURES = 'v047'
+
+
+{{ if .Env.GitReference }}
+[[groups]]
+  # warning: patch for `libp2p` uses the features mechanism. default-features and features will not take effect because the patch dependency does not support this mechanism
+  # warning: Patch `libp2p v0.47.0 (https://github.com/libp2p/rust-libp2p?branch=master#d0da3a09)` was not used in the crate graph.
+  id = "current"
+  instances = { count = 1 }
+
+  [groups.build_config.build_args]
+    CARGO_FEATURES = 'v047'
+    CARGO_REMOVE = 'libp2pv0470'
+    CARGO_PATCH = """
+libp2pv0470 = {package = "libp2p", git = "https://{{ or .Env.GitTarget "github.com/libp2p/rust-libp2p" }}", branch = "{{ .Env.GitReference }}", default_features = false, features = [ "websocket", "mplex", "yamux", "tcp-async-io", "ping", "noise", "dns-async-std" ], version = "0.47.0", optional = true}
+"""
+{{ end }}
```

Comment:
@mxinden I'd need your input on this,

I couldn't use dependencies overwrite because we use feature flags with libp2p,
so I resorted to `sed`'ing the current version and replacing it (see the corresponding Dockerfile where I use `CARGO_REMOVE`).

Can we leave with this? Are you aware of another approach?

Reply:
> I couldn't use dependencies overwrite because we use feature flags with libp2p,

I am not quite sure I am following. Are the feature flags (e.g. `v0470`) ignored when overriding a dependency?

As an aside, we can as well get rid of the above indirection:

``` diff
diff --git a/ping/rust/Cargo.toml b/ping/rust/Cargo.toml
index ac03ab5..5720ce7 100644
--- a/ping/rust/Cargo.toml
+++ b/ping/rust/Cargo.toml
@@ -3,12 +3,6 @@ edition = "2021"
 name = "testplan"
 version = "0.1.0"
 
-[features]
-v044 = ["libp2pv0440"]
-v045 = ["libp2pv0450"]
-v046 = ["libp2pv0460"]
-v047 = ["libp2pv0470"]
```

And instead use the package names as feature names directly, e.g. `libp2pv0470` as in `cargo build --release --features="libp2pv0470"`.

Reply:
👍 I'll drop the indirection,

> I am not quite sure I am following. Are the feature flags (e.g. v0470) ignored when overriding a dependency?

Yes, when I patch `libp2pv0470` without `features = [ "websocket", "mplex", ...]`, I get:

```
warning: Patch `libp2p v0.47.0 (https://github.com/laurentsenta/rust-libp2p?branch=demo/pass#f888503a)` was not used in the crate graph.
Check that the patched package version and available features are compatible with the dependency requirements.
```

If I pass the `features = ...` options, I get:

```
warning: patch for `libp2p` uses the features mechanism. default-features and features will not take effect because the patch dependency does not support this mechanism
```

Gist with the attempts and full logs:
https://gist.github.com/laurentsenta/e1ac5166d595245b8d278a4cab6caf96

Reply:
Just to double check, when you do the override, you override `libp2pv0470` AND you set the feature flag `libp2pv0470`, correct @laurentsenta?

Reply:
Yes, the cargo looks like this:
(attempt 2 in the gist)

```toml
[package]
edition = "2021"
name = "testplan"
version = "0.1.0"

[dependencies]
async-std = {version = "1.10", features = ["attributes", "tokio1"]}
env_logger = "0.9.0"
...
libp2pv0460 = {package = "libp2p", default_features = false, features = [ "websocket", "mplex", "yamux", "tcp-async-io", "ping", "noise", "dns-async-std" ], version = "0.46.0", optional = true}
libp2pv0470 = {package = "libp2p", git = "https://github.com/libp2p/rust-libp2p", branch = "master", default_features = false, features = [ "websocket", "mplex", "yamux", "tcp-async-io", "ping", "noise", "dns-async-std" ], version = "0.47.0", optional = true}

[patch.crates-io]
libp2pv0470 = { package = 'libp2p', git = 'https://github.com/laurentsenta/rust-libp2p', branch = 'demo/pass' }
```

If I run `cargo update && cargo build --features libp2pv0470`

I get the output:

> warning: Patch `libp2p v0.47.0 (https://github.com/laurentsenta/rust-libp2p?branch=demo/pass#f888503a)` was not used in the crate graph.
> Check that the patched package version and available features are compatible
> with the dependency requirements. If the patch has a different version from
> what is locked in the Cargo.lock file, run `cargo update` to use the new
> version. This may also occur with an optional dependency that is not enabled.
>     Finished dev [unoptimized + debuginfo] target(s) in 0.13s